src/heif.imageio/heifinput.cpp

```diff
@@ -0,0 +1,290 @@
+    uint8_t magic[magic_size];
+    FILE *file = Filesystem::fopen(filename, "rb");
+    fread (magic, magic_size, 1, file);
+    fclose (file);
+    heif_filetype_result filetype_check = heif_check_filetype(magic,12);
+    return filetype_check != heif_filetype_no
+            && filetype_check != heif_filetype_yes_unsupported
+    // This is what the libheif example said to do, but I can't find
+    // the filetype constants declared anywhere. Are they obsolete?
+}
+#endif
+
+
+
+bool
+HeifInput::open(const std::string& name, ImageSpec& newspec)
+{
+    // If user doesn't want to provide any config, just use an empty spec.
+    ImageSpec config;
+    return open(name, newspec, config);
+}
+
+
+
+bool
+HeifInput::open(const std::string& name, ImageSpec& newspec,
+                const ImageSpec& config)
+{
+    m_filename = name;
+    m_subimage = -1;
+
+    m_ctx.reset(new heif::Context);
+    m_himage  = heif::Image();
+    m_ihandle = heif::ImageHandle();
+
+    try {
+        m_ctx->read_from_file(name);
+        // FIXME: should someday be read_from_reader to give full flexibility
+
+        m_item_ids   = m_ctx->get_list_of_top_level_image_IDs();
+        m_primary_id = m_ctx->get_primary_image_ID();
+        for (size_t i = 0; i < m_item_ids.size(); ++i)
+            if (m_item_ids[i] == m_primary_id) {
+                m_item_ids.erase(m_item_ids.begin() + i);
+                break;
+            }
+        // std::cout << " primary id: " << m_primary_id << "\n";
+        // std::cout << " item ids: " << Strutil::join(m_item_ids, ", ") << "\n";
+        m_num_subimages = 1 + int(m_item_ids.size());
+
+    } catch (const heif::Error& err) {
+        std::string e = err.get_message();
+        errorf("%s", e.empty() ? "unknown exception" : e.c_str());
+        return false;
+    } catch (const std::exception& err) {
+        std::string e = err.what();
+        errorf("%s", e.empty() ? "unknown exception" : e.c_str());
+        return false;
+    }
+
+    bool ok = seek_subimage(0, 0);
+    newspec = spec();
+    return ok;
+}
+
+
+
+bool
+HeifInput::close()
+{
+    m_himage  = heif::Image();
+    m_ihandle = heif::ImageHandle();
+    m_ctx.reset();
+    m_subimage      = -1;
+    m_num_subimages = 0;
+    return true;
+}
+
+
+
+bool
+HeifInput::seek_subimage(int subimage, int miplevel)
+{
+    if (miplevel != 0)
+        return false;
+
+    if (subimage == m_subimage) {
+        return true;  // already there
+    }
+
+    if (subimage >= m_num_subimages) {
+        return false;
+    }
+
+    try {
+        auto id     = (subimage == 0) ? m_primary_id : m_item_ids[subimage - 1];
+        m_ihandle   = m_ctx->get_image_handle(id);
+        m_has_alpha = m_ihandle.has_alpha_channel();
+        auto chroma = m_has_alpha ? heif_chroma_interleaved_RGBA
+                                  : heif_chroma_interleaved_RGB;
+        m_himage = m_ihandle.decode_image(heif_colorspace_RGB, chroma);
+
+    } catch (const heif::Error& err) {
```

Comment:
Its a shame that the C++ wrapper for `libheif` converts error codes into exceptions and that you have to catch the exception to turn it back into an error code ...